api/src/internal/oscc.h

```diff
@@ -7,9 +7,32 @@
  #ifndef _OSCC_INTERNAL_H
  #define _OSCC_INTERNAL_H
 
+#include <net/if.h>
+#include <stdbool.h>
 
 #define CONSTRAIN(amt, low, high) ((amt) < (low) ? (low) : ((amt) > (high) ? (high) : (amt)))
 
+struct can_contains {
+    bool is_oscc;
+    bool has_vehicle;
+}can_contains_s;
+
+struct vehicle_can_desc {
+    bool has_steering_angle;
+    bool has_brake_pressure;
+    bool has_wheel_speed;
+}vehicle_can_desc_s;
+
+struct oscc_can_desc {
+    bool has_torqe_report;
+    bool has_steer_report;
+    bool has_brake_report;
+}oscc_can_desc_s;
+
+struct device_name {
+    struct device_name *next, *prev;
```

Comment:
Something else comes to mind - instead of using a linked list, can we just use an array of pointers to strings? Here's what I'm thinking:

```c
#define MAX_INTERFACES 32

size_t bufsize = sizeof( char * ) * MAX_INTERFACES;
char ** device_list = malloc( bufsize );
// do things
```

This could be packed in a struct to contain the size as well:

```c
struct device_list
{
    char **dev_names;
    size_t size;
}
```

Or, we could just allocate fixed sized buffers:

```c
struct device_list
{
    char[MAX_INTERFACES][IFNAMSIZ] devnames;
    size_t size;
}
```

While a linked list could work here, I'm not convinced that it's the right data structure. Since this isn't running on an embedded device and there's a reasonable bound on size of this data structure I think we could use a very large buffer size and assume that people aren't running the OSCC API on a Cisco network switch. (I may regret this comment later when Cisco comes out with a self driving datacenter rack.)

Reply:
I'd agree, choice of a linked list was more from lack of certainty on what to set for a MAX_INTERFACES. We most certainly could do that and it would make things easier. Part of this uncertainty is fear of devices like 5 port Ethernet cards and 4 port CAN devices that could multiply quickly if someone has a few of these installed on their machine, albeit somewhat extreme.  Maybe linux already has a max number... :thinking: